### ISSUE TITLE
Custom task list token in Visual Studio 2022 doesn't work for Python.

### ISSUE THREAD

**Issue by vsfeedback** [microsoft/PTVS#7151]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Custom-task-list-token-in-Visual-Studio-/10133217)._

---
Hi Team,

I found that my custom task list tokens don't work in VS 2022 for Python(But it works in VS 2019).

**Steps to reproduce:**

1. Open VS 2022, create a Python project.
2. In Tools > Options > Environment > Task List > input a new name for example "Test1" > click Add to add this custom task list token.
3. In .py file > input some comments for example "#Test1: this is a test" > save the project.
4. Click View > Task List.
5. In Task List window, the Test1 task list token doesn't appear.

You can test in C# projects, or in VS 2019 Python projects, they should work well.

If it's a potential issue, please fix it.

Thanks

---
#### Feedback Bot on 8/29/2022, 02:23 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by K-E-K** [comment]
I am having the same problem but with even the default TODO tag, though it's working normally in VS 2019.

**Comment by kue43** [comment]
Visual Studio 2022 task list not working at all for Python

**Comment by janestebans** [comment]
C++, not working for me either. I have the Options->TextEditor->C/C++->View->Enumerate Comment Tasks = true. Tried switching it off and on and restarte VS, did not work.
It is wierd because it shows the comments from boost library files but not mine. Adding a new task list token does not work either.

- Microsoft Visual Studio Community 2022 (64-bit) 
- VS version: 17.3.6
- Windows 11.

Thanks.

**Comment by AdamYoblick** [comment]
Thanks for the additional info. I was going to say that the C++ language server is different from the python one, but it sounds like you found a workaround 😄 

On the python side, I'm have a proof of concept working for the task list. I needed to add a feature to pylance (the language server). Now I just need to flesh it out.

**Comment by K-E-K** [comment]
I have upgraded to : Version 17.4.1
(Microsoft Visual Studio Professional 2022 ) and the task list is still not working at all for Python.

**Comment by AdamYoblick** [comment]
Yep, I'm still working on this. I get pulled into infra work and other higher priority tasks often so my work gets interrupted. I'm hoping to complete the work this week if all goes well. I'll update this issue when the fix is in.

**Comment by K-E-K** [comment]
Ok glad to hear that. Thank you for keeping us updated.

**Comment by AdamYoblick** [comment]
This is now working for me locally, I'll reply when I know what version of Visual Studio this will be released in.

**Comment by pbarranis** [comment]
@AdamYoblick I have found some different but similar issues with the Task List since upgrading to 17.4.  I'll describe briefly - you tell me if this is close enough to the issues you are fixing that it does or doesn't warrant opening a separate ticket, please.

I have two solutions I work on routinely, both primarily C# with some TS/HTML/SASS.  In one solution, the Task List went totally blank after upgrading.  Now that I'm looking more closely, I see that even the column headings aren't visible.  

Tools -> Options -> Environment -> Task List shows my usual, heavily-customized list, and there should be dozens of entries in the Task List at varying priorities.

In the other solution, all of my task list items show up.  However, all the items that should be high & low priority show up as normal priority, and all the items that should be normal priority show up as low priority.

Per a post on SO for whom this worked, I tried doing a Repair on VS through the Installer.  Unfortunately, no joy.

TIA.  I can open a new ticket if needed.  Excited to have this potentially fixed soon, as obviously I'm a bit of a Task List power user.

**Comment by randsu** [comment]
Task List for Visual Studio Professional 2022 (64 bit) version 17.4.3
does not pick up // TODO in code.

**Comment by AdamYoblick** [comment]
> @AdamYoblick Adam Yoblick FTE I have found some different but similar issues with the Task List since upgrading to 17.4. I'll describe briefly - you tell me if this is close enough to the issues you are fixing that it does or doesn't warrant opening a separate ticket, please.
> 
> I have two solutions I work on routinely, both primarily C# with some TS/HTML/SASS. In one solution, the Task List went totally blank after upgrading. Now that I'm looking more closely, I see that even the column headings aren't visible.
> 
> Tools -> Options -> Environment -> Task List shows my usual, heavily-customized list, and there should be dozens of entries in the Task List at varying priorities.
> 
> In the other solution, all of my task list items show up. However, all the items that should be high & low priority show up as normal priority, and all the items that should be normal priority show up as low priority.
> 
> Per a post on SO for whom this worked, I tried doing a Repair on VS through the Installer. Unfortunately, no joy.
> 
> TIA. I can open a new ticket if needed. Excited to have this potentially fixed soon, as obviously I'm a bit of a Task List power user.

Sorry for the delay, I've been on vacation for the past month. Task list diagnostics are handled by each language team's language server, so any fixes I make to the python one won't trickle down to the other languages. I will reach out to the visual studio team to see if they can re-route this request for the other languages as well.

For python, my changes are in code review and will be committed in the next day or so. But it will take time to make it into a visual studio release. I'll reply back once I know what version it will be in.

**Comment by AbhimanyuVS** [comment]
Hello,

I have tried all of the different solutions listed in this thread, but I am still unable to see any tasks (specified by either custom or default tokens) in VS2022. It works in VS2019, for Python, C++, and C#. However, it does not work in VS2022, for any language.

For reference, I posted a bug report about this here: https://developercommunity.visualstudio.com/t/Task-list-is-not-working-in-VS2022/10192598

This was then piped here: https://github.com/microsoft/PTVS/issues/7254

That was then closed citing this thread as a duplicate.

I look forward to a solution to this problem, as both VS2022 and the task list are very nice.

Thank you.

**Comment by smimon** [comment]
I have been experiencing blank task list in VS 2022 with C# for months.

Doing a full repair fixes them temporarily - close down and re-open the solution a few times, they all disappear again.

Although internally to the MS team working on this, this may not be considered high priority, this is a fairly critical piece of productivity for a lot of developers, so may we please request that this is flagged as such and escalated for a quicker fix across all languages?

**Comment by AdamYoblick** [comment]
This is working locally with changes to both pylance and the VS Editor:

![image](https://user-images.githubusercontent.com/29436557/217396824-d5500ab2-1741-4f4c-b942-a4afa7b8aa2f.png)

I still need to go through the VS PR process, etc, but good progress 😄

**Comment by AdamYoblick** [comment]
Note that task list work with just pylance changes, but priority requires VS changes. I'm going to get the task list PR in ASAP, then work on getting https://github.com/microsoft/PTVS/issues/7300 in.

**Comment by AdamYoblick** [comment]
This has been merged in and will be in the next release public release of VS 2022.

**Comment by AbhimanyuVS** [comment]
@AdamYoblick 

I installed update 17.4.5, but I still cannot see the task list in Python projects. Will this issue be addressed in another update? Sorry if I misunderstood your last comment.

**Comment by AdamYoblick** [comment]
@AbhimanyuVS This change just got merged into pylance, but there's a sequence of releases that have to happen first. Unfortunately, that means we can't just release bugfixes whenever we want.

Pylance has to release first, which happens once per week. Once there's a new version of pylance, PTVS can consume it and PTVS can insert into Visual Studio. Then Visual Studio needs to release. The GA (general availability) release of Visual Studio 2022 17.6 is around May 9th, 2023.

I know that's not what you wanted to hear, but since we insert into Visual Studio, we release with them as well. 😄

**Comment by AbhimanyuVS** [comment]
@AdamYoblick 

Understood, not a problem. Now that I know the ETA, I am alright with waiting for it. Thank you for your work on the same; I look forward to receiving the fix in May.

Since this is closed, should I create separate tickets for (a) the task list not working for C++ and (b) the task list behaving strangely with C#?

**Comment by AdamYoblick** [comment]
@AbhimanyuVS I tested C# myself and it works for me locally, which means it will be fixed in the next release as well. But if you want to create a new issue, you definitely can. C# development is done at 
https://github.com/dotnet/roslyn

I'm not sure about C++ though. I think you'd need to create a feedback ticket through Visual Studio and it will get routed to the correct team as long as you specify C++ in the title. 😄

**Comment by AbhimanyuVS** [comment]
@AdamYoblick

Good to know that the task list will work with C#. Thank you for pointing me to the roslyn repo.

I will put in a new ticket for the task list in C++.

Thank you for your time and consideration.

**Comment by CaligoLab** [comment]
Just for the record:
VS Version 17.4.4 (Win x64) up to date.
This is my first try at creating tasks (Python)
Neither TODO, nor custom markers are picked up - the task list is empty.

**Comment by AbhimanyuVS** [comment]
For reasons unknown, this issue has resurfaced for me. In Oct 2024, the task list was working fine for Python source code in Visual Studio 2022. However, at some point since then, it has stopped working. Tasks no longer show up for me.

**Comment by ScatteredDesign** [comment]
Yeah, still not working in C# projects. Hugely annoying....

**Comment by donr484** [comment]
UPDATE...

After messing with it for a while, I tried closing the solution and reopening. 
That got the todo comments to show up in the task list again for c#!

Another issue I see is that I cannot add tokens to that task list token list under Tools -> Options.
That problem did not go away when I closed and reopened the solution.

Being able to add the toggle task list bookmark to a toolbar would still be nice too.
If we could right-click a line and toggle a task list bookmark, that would be a decent alternative.

-------

I am having the same issue for C# projects in version 17.14.8. (It works fine for VB.NET projects.)

Toggling a task list bookmark will get a comment to show up, but there doesn't seem to be a way to add that to the toolbar. I checked the Edit commands and many other ones that could be relevant. Where is that command?

The toolbar has a toggle bookmark icon, but it doesn't work with the task list.

There is no way to remove a task list bookmark from the task list or with a click on the line in the ide.

The only way to use this feature is a clunky two part keyboard shortcut or a nested dropdown menu.

It is pretty frustrating.
